File: chap12/fig12-12.R
library(mvtnorm)
library(dplyr)
library(ggplot2)

set.seed(123)

N <- 101
X <- seq(0, 1, len=N)
S <- 5

a_s   <- c(2, 0.5)
rho_s <- c(0.05, 0.2)
params <- tidyr::crossing(a=a_s, rho=rho_s)

d <- lapply(1:nrow(params), function(paraID) {
  a   <- params$a[paraID]
  rho <- params$rho[paraID]
  K <- matrix(rep(0, N*N), nrow=N)
  for (i in 1:N) {
    for (j in 1:N) {
      K[i,j] <- a^2 * exp(-0.5/(rho^2)*(X[i]-X[j])^2)
    }
  }
  ys <- sapply(1:S, function(s) {
    y <- rmvnorm(1, mean=rep(0, N), sigma=K)
  })
  data.frame(a=a, rho=rho, X=X, ys) %>% 
    magrittr::set_colnames(c('a', 'rho', 'X', 1:S)) %>% 
    tidyr::pivot_longer(cols=-c(a, rho, X), names_to='Sample', values_to='Y')
}) %>% bind_rows()

d_plot <- d %>% 
  mutate(a_disp=paste0('a==', a), rho_disp=paste0('rho==', rho))

p <- ggplot(data=d_plot, aes(x=X, y=Y, linetype=Sample)) +
  theme_bw(base_size=18) +
  theme(axis.text.x=element_text(angle=40, vjust=1, hjust=1)) +
  facet_grid(rho_disp ~ a_disp, labeller=label_parsed) +
  geom_line(alpha=0.8)
ggsave(p, file='output/fig12-12.png', dpi=300, w=6, h=5)

File: chap12/fig12-12.py
import pandas
import itertools
import numpy as np
from plotnine import ggplot, aes, theme_bw, theme, element_text, facet_grid, geom_line

np.random.seed(123)

N = 101
X = np.linspace(0, 1, num=N)
S = 5

a_s = [2, 0.5]
rho_s = [0.05, 0.2]
d_plot = pandas.DataFrame(index=[], columns=['a', 'rho', 'X', 'Sample', 'Y'])
for a, rho in itertools.product(a_s, rho_s):    
    K = np.zeros((N, N))
    for i in range(N):
        for j in range(N):
            K[i,j] = np.square(a) * np.exp(-0.5/np.square(rho)*np.square(X[i]-X[j]))
    ys = np.random.multivariate_normal(np.repeat(0,N), K, S)
    d_ = pandas.DataFrame(np.column_stack([np.repeat(a, N), np.repeat(rho, N), X, ys.T]), \
        columns=['a', 'rho', 'X', *list(range(S))])
    d_ = pandas.melt(d_, id_vars=['a', 'rho', 'X'], var_name='Sample', value_name='Y')
    d_['Sample'] = pandas.Categorical(d_.Sample, categories=list(range(S)))
    d_plot = pandas.concat([d_plot, d_])

p = (ggplot(d_plot, aes('X', 'Y', group='Sample'))
    + theme_bw(base_size=18)
    + theme(axis_text_x=element_text(angle=40, vjust=1, hjust=1))
    + facet_grid('rho ~ a')
    + geom_line(alpha=0.8)
)
p.save(filename='output/fig12-12.py.png', dpi=300, width=6, height=5)
